src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,108 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check, and to declare that a programmer has checked that
+/// these contracts have been upheld.
+///
+/// Unsafe Rust add some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, compiler intrinsics,
+/// and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # Dereferencing raw pointers
+///
+/// ```rust
+/// let a = 3;
+/// let b = &a as *const _;
+/// // SAFETY: we know that b will always be a valid address.
+/// let _ = unsafe { *b };
+/// ```
+///
+/// # Marking elements as unsafe
+///
+/// `unsafe` can be used on functions. Note that functions declared in
+/// [`extern`] blocks are implicitly marked as `unsafe`. Methods can also be
+/// declared as `unsafe`:
+///
+/// ```rust
+/// #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// extern "C" {
+///     fn unsafe_extern_fn();
+/// }
+///
+/// trait TraitWithUnsafeMethod {
+///     unsafe fn unsafe_method();
+/// }
+///
+/// struct S;
+///
+/// impl S {
+///     unsafe fn unsafe_method_on_struct() {}
+/// }
+/// ```
+///
+/// Traits can also be declared as `unsafe`:
+///
+/// ```rust
+/// unsafe trait UnsafeTrait {}
+/// ```
+///
+/// # Using `unsafe` blocks and `impl`s
+///
+/// Calling an `unsafe` function requires an `unsafe` block, which can take
+/// two forms:
+///
+/// ```rust
+/// #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// // This function is unsafe to call and unsafe functions can be called
+/// // inside without problems.
+/// unsafe fn enclosing_unsafe() {
+///     unsafe_fn()
+/// }
+///
+/// // This function is safe to call. Any call to an unsafe function inside
+/// // must be enclosed in an unsafe block.
+/// fn interior_unsafe() {
+///     // SAFETY: explanation about why this block is safe in context.
+///     unsafe { unsafe_fn() }
+/// }
+/// ```
 ///
+/// Traits marked as `unsafe` must be [`impl`]emented using `unsafe`:
```

Comment:
```suggestion
/// Traits marked as `unsafe` must be [`impl`]emented using `unsafe`.
/// This makes a guarantee to other unsafe code that a property holds for this type.
/// It is undefined behavior to make an `unsafe impl` for a type which does not have this property.
```
Might also be nice to use `Send` and `Sync` as an example here.

Reply:
I added them, testing my changes locally before pushing right now 😄  !